Codeclimate improvements

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

* Gets `rubocop` run green.
* ~Gets `bundle-audit` green by upgrading `nokogiri` to a non vulnerable version~. (Already landed in through #1306).
* Brings back ruby code style checks in `codeclimate`.
* Fix file exclusions so that `codeclimate` can be run locally in a reasonable amount of time.

I think all offense corrections should be safe but please review carefully... Also, I wasn't sure about
https://github.com/deivid-rodriguez/decidim/commit/06d381a54ceb69a3b9b90cadd621117af9b0c8c1. That method name is duplicated in the class, you probably know best if it's actually needed or if this is fine. Tests seem to pass without it (an increasing code coverage percentage would also be an sign that it was a dead method).

#### :ghost: GIF

![cat](https://cloud.githubusercontent.com/assets/2887858/25435205/a52469ea-2a65-11e7-98ea-a73de8945529.gif)

oriolgual commented:
@deivid-rodriguez can you rebase against master to add the proper CircleCI config?

deivid-rodriguez commented:
It should be rebased.

oriolgual commented:
@deivid-rodriguez it was an issue with Circle's configuration, can you rebase again & push to trigger a build?

deivid-rodriguez commented:
The coverage decrease seems to be due to the new environment variable set in circle, `SIMPLECOV`. It was introduced to replace the `CI` variable in one place, but not everywhere. So my guess is that Travis is sending "part of the coverage" where allowed by the leftovers of the `CI` variable?

deivid-rodriguez commented:
> @deivid-rodriguez it was an issue with Circle's configuration, can you rebase again & push to trigger a build?

Of course! :wink:

deivid-rodriguez commented:
Done!

deivid-rodriguez commented:
Seems like CircleCI picked up the build this time! :tada: Now it can have a proper race with Travis :stuck_out_tongue_closed_eyes:.

mrcasals commented on decidim-core/app/helpers/decidim/widget_urls_helper.rb:
```diff
@@ -2,8 +2,8 @@
     def embed_modal_for(url)
-      js_embed_code = "#{content_tag(:script, '', src: url)}"
-      embed_code = "#{content_tag(:noscript, content_tag(:iframe, '', src: url.gsub(".js", ".html"), frameborder: 0, scrolling: "vertical"))}"
+      js_embed_code = content_tag(:script, "", src: url).to_s
```
We need to check this manually, I guess

deivid-rodriguez replied:
If I understand it correctly, string interpolation just implicitly calls `to_s` on every interpolated expression and the concatenates everything. So interpolating a single expression is equivalent to just calling `to_s` on it.

However, the auto-correction cop might be missing some subtlety and could have a bug, of course. So it never hurts to check this manually.

mrcasals replied:
Yes, what I'm fearing is that `content_tag` is directly generating HTML somehow, but we need to save it to a string so that we can render it as is (not HTML safe, I mean). To be fair I missed the `to_s` the first time I saw the changes and I think it should work, but we'll add a review app somewhere and check all these changes :)

beagleknight replied:
Hey @deivid-rodriguez . We tried `to_str` instead of `to_s` and it works like a charm. Can you change it please? 😄

deivid-rodriguez replied:
You mean `to_s` didn't work? That's very surprising!

deivid-rodriguez replied:
Strictly speaking the 100% equivalent behavior would be using `to_s`, that's why I'm surprised. 

```
[1] pry(main)> class A
[1] pry(main)*   def to_s
[1] pry(main)*     "hola"
[1] pry(main)*   end  
[1] pry(main)*   
[1] pry(main)*   def to_str
[1] pry(main)*     "hi"
[1] pry(main)*   end  
[1] pry(main)* end  
=> :to_str
[2] pry(main)> "#{A.new}"
=> "hola"
[3] pry(main)> 
```

Also note that if the result of `content_tag` is already a `String` instance, I'd say you don't need either.

mrcasals replied:
Yes, it's weird but "content_tag" returns a SafeBuffer, which has the "to_s" method overwritten to return self. In order to test this, visit a participatory process in the development app and try to embed it (on the sidebar). If the code works fine, you'll be able to copy a raw HTML code. If it doesn't, you'll see some weird layout right after the page footer (that's how we checked it)

deivid-rodriguez replied:
That was a nice catch! So we just learned that this RuboCop auto-correction is not 100% safe. I'm happy I didn't break all your installations :stuck_out_tongue:...

So should I change this to `to_str`? I wonder if it's better to leave it as it is and disable the cop inline explaining the issue with a comment, so that nobody in the future will attemp this kind of change again.

mrcasals replied:
While I agree with leaving a comment here explaining what's going on (this is a non-common use of a Rails feature, I guess, I'm aware of that), I'd change the code to use `to_str`. What we used to have is really ugly (`"#{content_tag(...)}"`), `to_str` looks cleaner to me.

What do you think @deivid-rodriguez? :)

deivid-rodriguez replied:
Sure, sounds good! I'm just not sure of the wording.

By the way, `String.new(content_tag(...))` works as well, and it blows up if `to_str` is not defined or does not return a `String` instance. So maybe it's a bit more resilient than `to_str`?

```
[1] pry(main)> class A
[1] pry(main)*   def to_str
[1] pry(main)*     "hi"
[1] pry(main)*   end  
[1] pry(main)* end  
=> :to_str
[2] pry(main)> class B
[2] pry(main)*   def to_str
[2] pry(main)*     self
[2] pry(main)*   end  
[2] pry(main)* end  
=> :to_str
[3] pry(main)> class C; end
=> nil
[4] pry(main)> String.new(A.new)
=> "hi"
[5] pry(main)> String.new(B.new)
TypeError: can't convert B to String (B#to_str gives B)
from (pry):13:in `initialize'
[6] pry(main)> String.new(C.new)
TypeError: no implicit conversion of C into String
from (pry):14:in `initialize'
```

The lesson here, I think, is that `to_s` seems to be good for "object representation" and `to_str` for "type conversion", and we want the latter here.

Sorry for disgressing this much, at least we are learning Ruby :)

mrcasals replied:
We tried `String.new` too, but we went for `to_str` because it was shorter 😆 But it might lead to confusion, someone might think it's a typo and change it to `to_s` which will break things, so `String.new` wins here (apart from being more resilient as you pointed out). So go for it and please add a comment explaining why we need to turn this into a String (so that it doesn't render the HTML).

Another option would be writing the raw HTML instead of using `content_for`, but seems more error-prone to me 😕

deivid-rodriguez replied:
Ok, so 4ae83e510b7ad342c54ef8f019f981c9f7690770 is my proposal. Instead of a comment, I added a regression test that serves the same purpose but in a nicer way, I think!

mrcasals commented:
Actually, we need to check some things after this PR, but it looks good to me. CircleCI has a green test suite, let's wait for TravisCI and see what does it report.

anyway, thanks for this awesome PR! <3 🎉

deivid-rodriguez commented:
You're welcome!

Travis is green too. Also, the coverage for [decidim-core/app/controllers/decidim/authorizations_controller.rb](https://github.com/deivid-rodriguez/decidim/blob/6e19076ed6c7d6a3215b45d13970b398ae6db023/decidim-core/app/controllers/decidim/authorizations_controller.rb) increased so it's more likely that the method removed was really dead.

Rebasing again due to conflict with master.

deivid-rodriguez commented on decidim-admin/app/commands/decidim/admin/create_attachment.rb:
```diff
@@ -28,7 +28,7 @@ def call
           broadcast(:ok)
         else
-          if @attachment.errors.has_key? :file
+          if @attachment.errors.key? :file
```
I actually like `has_key?` better here. This cop allows configuring it for `is_`, `has_` or `have_` methods, so I could whitelist `has_` methods if you agree.

oriolgual replied:
+1 with has_key? !

deivid-rodriguez replied:
It seems that `key?` (3 instances) is actually more common than `has_key?` (1 instance) in the codebase, so maybe others have different opinion...

Rubocop hash a specific cop to check for "hash methods style", so the configuration leading to the smallest diff would be to allow methods starting with `has_` but explicitly prefer `keys?` to `has_keys?`.

oriolgual replied:
I think there are smaller number of instances because we ran `rubocop -a` and autocorrected it in the past, but no because it was preferred.

deivid-rodriguez replied:
I see. I'll update the PR then to allow `has_` methods as proposed and change `keys?` usages to `has_keys?`.

deivid-rodriguez commented:
Rebased the PR and added one more improvement: https://github.com/decidim/decidim/pull/1290/commits/e0afd9d9903d3bb5700d41661c2e438305979a23. I wanted to get the codecov report green. Didn't succeed but the patch seems good to me!

deivid-rodriguez commented:
This PR's size was getting out of hand, so I removed the latest changes and move them to separate PR's (#1314 and #1315).

mrcasals commented:
Codecov seems to be hung or something, but the PR looks good to me, so I'm 👍 for merging this. Let's wait for someone else to 👍 and merge this :)

mrcasals commented on decidim-core/lib/decidim/form_builder.rb:
```diff
@@ -20,6 +20,7 @@ class FormBuilder < FoundationRailsHelper::FormBuilder
     # html_options - a Hash with options
     #
     # Renders a collection of check boxes.
+    # rubocop:disable Metrics/ParameterLists
```
I really hate having to do this (I understand we can't help it since we're overriding a method, but...)
